src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -4451,6 +4451,76 @@ public override IOperation VisitMemberInitializer(IMemberInitializerOperation op
             throw ExceptionUtilities.Unreachable;
         }
 
+        public override IOperation VisitArrayCreation(IArrayCreationOperation operation, int? captureIdForResult)
+        {
+            // We have couple of options on how to rewrite an array creation with an initializer:
+            //       1) Retain the original tree shape so the visited IArrayCreationOperation still has an IArrayInitializerOperation child node.
+            //       2) Lower the IArrayCreationOperation so it always has a null initializer, followed by explicit assignments
+            //          of the form "IArrayElementReference = value" for the array initializer values.
+            //          There will be no IArrayInitializerOperation in the tree with approach.
+            //
+            //  We are going ahead with approach #1 for couple of reasons:
+            //  1. Simplicity: The implementation is much simpler, and has a lot lower risk associated with it.
+            //  2. Lack of array instance access in the initializer: Unlike the object/collection initializer scenario,
+            //     where the initializer can access the instance being initialized, array initializer does not have access
+            //     to the array instance being initialized, and hence it does not matter if the array allocation is done
+            //     before visiting the initializers or not.
```

Comment:
> While it is technically true that side affects shouldn't be observable to the initializers of an array, this doesn't accurately reflect the code that is actually generated by the compiler. I think we should probably revisit this sooner rather than later, especially as some array initializers actually invoke methods to set elements (multidimension arrays call arraytype::Set, for example, not array assignment).

I actually think the implemented approach is the right approach for CFG. The reasons are outlined in the comment and we wouldn't lower the nodes to arraytype::Set calls anyway.

---
In reply to: [188776536](https://github.com/dotnet/roslyn/pull/26901#discussion_r188776536) [](ancestors = 188776536)